3.4 NOCHANGE: hangprinter: simplify things by turning unrolled stuff into loops

jtimon opened the pull request:
This is in preparation of potentially supporting more anchors for different hangprinter setups than the "standard" one, (4 anchors, 3 base anchors plus one on top for stability).
pinging @tobbelobb to have an opinion on the changes if possible.

There is no functional change in this PR (unless it has bugs).

tobbelobb commented on src/Movement/Kinematics/HangprinterKinematics.cpp:
```diff
@@ -275,22 +274,15 @@ bool HangprinterKinematics::CartesianToMotorSteps(const float machinePos[], cons
 													size_t numVisibleAxes, size_t numTotalAxes, int32_t motorPos[], bool isCoordinated) const noexcept
 {
 	float squaredLineLengths[HANGPRINTER_AXES];
-	squaredLineLengths[A_AXIS] = LineLengthSquared(machinePos, anchors[A_AXIS]);
```
I would prefer three separate for-loops. It keeps separate calculations separate which is good from a math reasoning perspective. It might also be good from a processor cache/optimization perspective.

jtimon replied:
I don't prefer that. But, as you wish.
I mean, it's a style thing we could only objectively resolve by looking at assembly code, and I don't feel like to.
Your caching hypothesis is reasonable and I'm too lazy to try to refute it.
I didn't like doing this, but done.
The bike shed color is yours.

tobbelobb commented:
I like this change. The unrolls were only a coincidence, I had no reason to unroll. I like all changes that bring us closer to Hangprinter's number of axes being a config.g variable.

I'd like to see a test on hardware or at least some kind of typo-safeguarding test before merge.

jtimon committed:
Hangprinter: Loop around the anchors instead of naming them directly

jtimon committed:
HangprinterKinematics::WriteCalibrationParameters: flat return style

Otherwise it gets too deep.
I think there's an option in eclipse to forbid this by project in
format, one can config depness iirc.

jtimon commented on src/Movement/Kinematics/HangprinterKinematics.cpp:
```diff
@@ -107,16 +107,15 @@ void HangprinterKinematics::Recalc() noexcept
 	float stepsPerUnitTimesRTmp[HANGPRINTER_AXES] = { 0.0 };
 	Platform& platform = reprap.GetPlatform(); // No const because we want to set drive steper per unit
-	for (size_t i = 0; i < HANGPRINTER_AXES; i++)
+	for (size_t i = 0; i < HANGPRINTER_AXES; ++i)
```
I know this diff line is unnecessary, it's just for consistency within the file.

jtimon commented:
I pushed a new version with improvements from the feedback and testing.

I tested M669 and M666 without parameters, but I still didn't manage to test HangprinterKinematics::WriteCalibrationParameters, how do I test that?

I also didn't test any of the modified parts related to CAN/Odrive. And I won't manage to test anytime soon, but at the same time those changes seem quite simple and inoffensive. 

Testers on that front are very welcomed.  For this PR, you just need a working hangprinter, upgrade to this build, test everything you normally do keeps working, report and go back to your previous working .bin, just in case.

Some parts can be tested without a hangprinter too, but we really don't want to break any of the current working hangprinters before we add any new experimental functionality.

tobbelobb commented:
> HangprinterKinematics::WriteCalibrationParameters, how do I test that?

I think ran `M500` and inspected the resulting `sys/config-override.g` manually. You could also do a `M501` to make the machine try to read and load the stored parameters. Remember to delete `sys/config-override.g` when done.

> Testers on that front are very welcomed. For this PR, you just need a working hangprinter, upgrade to this build, test everything you normally do keeps working, report and go back to your previous working .bin, just in case.

I'll see if I can put it on my machine within a couple of weeks. Will report back here.

jtimon commented:
Thank you. But, actually, perhaps for your testing you should wait for the rebased version of this PR, since more things changed than I expected.  Some of the changes here may no longer needed. I will post a reference here to the new one and close this one when it's done.

jtimon commented:
Closing in favor of https://github.com/Duet3D/RepRapFirmware/pull/597
